src/blockchain/new-block-emitters.ts

```diff
@@ -0,0 +1,129 @@
+import { BlockHeader, Eth } from 'web3-eth'
+import { loggingFactory } from '../logger'
+import { Subscription } from 'web3-core-subscriptions'
+import { EventEmitter } from 'events'
+import { Logger } from '../definitions'
+
+const DEFAULT_POLLING_INTERVAL = 5000
+export const NEW_BLOCK_EVENT_NAME = 'newBlock'
+
+/**
+ * Abstract EventEmitter that automatically start (what ever task defined in abstract start() method) when first listener is
+ * attached and similarly stops (what ever task defined in abstract stop() method) when last listener is removed.
+ */
+export abstract class AutoStartStopEventEmitter extends EventEmitter {
+  /**
+   * Name of event that triggers the start/stop actions. Eq. waits there is listeners for this specific event.
+   */
+  private readonly triggerEventName: string
+  protected logger: Logger
+
+  protected constructor (logger: Logger, triggerEventName: string) {
+    super()
+    this.logger = logger
```

Comment:
Do you need to bind the logger? Can it not live as a global variable or is the idea that the `start` and `stop` handlers can access it from `this`? Did you consider passing it as argument instead?

Reply:
This is because it is a base class for multiple different classes and you would get very unclear logs. Eq. you would not be able to distinguish what classes started/stopped listening.